### PR TITLE
fix_: make member use wallet tokens during permission checking

### LINKED ISSUE #14913: Members should use the wallet tokens instead of checking the chain when checking permissions
### Description

Currently, when members need to check if they fit a permission for the edit shared address popup or the channel blurred screen, we do calls to providers to know on chain if we have the correct tokens.

However, this is very wasteful, because the wallet already has that information. We should be using the standard wallet APIs to ask if we the tokens.

If we are concerned about the accuracy of the data in the wallet, we can call the function to refresh the balances, but IMO it shouldn't be necessary. The wallet already tracks the activity and should update over time.

### Acceptance criteria

- the channel and join permission screens work as before, same with the shared addresses popup. The UI/UX should't be affected at all. If anything, it should be faster
- members no longer call the providers

### PR DESCRIPTION
**Pull request by kounkou**
### Description

During the checking of permissions that happens either on edit shared addresses or during the blurred screen when joining a community, the member will perform some permissions checking. Currently the permission checking is done by fetching tokens from the chain which is wasteful in terms of resources. These changes aims at using the cached tokens from the wallet which improves the speed of tokens retrieval and improve the user experience.

fixes #14913

### Acceptance criteria

- [x] the channel and join permission screens work as before, same with the shared addresses popup. The UI/UX should't be affected at all. If anything, it should be faster
- [x] members no longer call the providers
- [ ] Ensure that no integration breaking changes `In progress`

### Important changes:
- [x] Add a switch to differentiate fetching token by member and owner
- [x] Add API to fetch tokens from the wallet db

Closes https://github.com/status-im/status-desktop/issues/14913

### Manual testing

Please notince how fast the evaluation is compared to before

### Before 

[Screencast from 2024-06-02 01-17-01.webm](https://github.com/status-im/status-go/assets/2589171/b8705900-9697-41f4-97a7-70a84fb39865)

### After

[Screencast from 2024-06-02 01-59-40.webm](https://github.com/status-im/status-go/assets/2589171/33f75e36-a43a-4758-8456-768551d1675c)

### CODE REVIEW

File: protocol/communities/permission_checker.go
Comment (jrainville) [inline protocol/communities/permission_checker.go]: Since all of these lines are the same for both functions, could we move them to `fetchOwnedERC721Tokens`?

File: services/wallet/api.go
Comment (jrainville) [inline services/wallet/api.go]: Do we need to call this from the client?

**Review by jrainville — COMMENTED** [review]
Great job! Just small comments.
I'll also wait for the others to review since they know this code better.

FYI, @osmaczko did [a PR](https://github.com/status-im/status-go/pull/5256) also on that code. Hopefully it's doesn't conflict too much.
I don't think you need to also implement the `getOwnersForCollection` function from cache, since it's only for control nodes, but it's worth noting.

File: services/wallet/api.go
Comment (kounkou) [reply services/wallet/api.go]: Hum yes..., it's somehow `polymorphically` called from the `GetOwnedERC721Tokens` in order to retrieve the tokens by the member.
Please clarify if I didn't understand the question.

File: protocol/communities_messenger_helpers_test.go
Comment (mprakhov) [inline protocol/communities_messenger_helpers_test.go:75]: can we escape this?

Comment (mprakhov) [reply protocol/communities_messenger_helpers_test.go:75]: ok, it seems it we have other places with the sleep in this file

Comment (kounkou) [reply protocol/communities_messenger_helpers_test.go:75]: Right, I think it's safe to keep it.

File: services/wallet/collectibles/ownership_db.go
Comment (mprakhov) [inline services/wallet/collectibles/ownership_db.go:425]: could you please add a test for this fn... what should we get, when info is not found?

File: protocol/communities/manager.go
Comment (mprakhov) [inline protocol/communities/manager.go:320]: please, rebase on the latest develop branch. I think you'll not pass due to code coverage

Comment (mprakhov) [reply protocol/communities/manager.go:320]: 60% looks like you are ok

**Review by mprakhov — APPROVED** [review]
LGTM

File: protocol/communities/manager.go
Comment (kounkou) [reply protocol/communities/manager.go:320]: rebased and code climate looks good.

**Review by osmaczko — APPROVED** [review]
Looks good :+1:

**Comment by osmaczko** [comment]
> FYI, @osmaczko did https://github.com/status-im/status-go/pull/5256 also on that code. Hopefully it's doesn't conflict too much.

It conflicts pretty heavily as far as I can see :cry:

File: services/wallet/api.go
Comment (jrainville) [reply services/wallet/api.go]: Since it's in api.go, it means it can be called by clients (Nim). Is that necessary?

File: services/wallet/collectibles/ownership_db.go
Comment (kounkou) [reply services/wallet/collectibles/ownership_db.go:425]: sure, please note that this is exactly the same behavior as for `FetchBalancesByOwnerAndContractAddress`. Instead of looking on chain we just get the data in cache, but let me have a look if I can add a test that makes sense if not already implemented somewhere.

File: services/wallet/api.go
Comment (kounkou) [reply services/wallet/api.go]: > Since it's in api.go, it means it can be called by clients (Nim). Is that necessary?

EDIT : 

I thought about adding a param. However even if I implement adding a parameter for the type of Fetching... (onChain or cache), it means the client can still  call the the cached version by just adding a parameter. Which doesn't resolve the core of your question. The reality is that collectibles constitute another service (wallet). Right now I don't know of any other way to reach the `FetchBalancesByOwnerAndContractAddress` other than going through the api.

**Comment by anastasiyaig** [comment]
@kounkou hey, is it the change you wanted to be tested? If so, please create corresponding PR in desktop repo

**Comment by kounkou** [comment]
> @kounkou hey, is it the change you wanted to be tested? If so, please create corresponding PR in desktop repo

Hey @anastasiyaig please let me follow up with you once I have completed the merging of changes from @osmaczko PR. Should be achieve early next week.

PR : https://github.com/status-im/status-desktop/pull/15117

**Comment by kounkou** [comment]
During the testing, @glitchminer saw a problem, but it's actually not caused by this pr, it is already known issue when trying to use the same collectible multiple times when creating a permission :

https://github.com/status-im/status-desktop/issues/15122

**Pull request merged**
